evncr.1
.Dd $Mdocdate$
.Dt evncr 1
.Os
.Sh NAME
.Nm evncr
.Nd read stdin in the voice of EVAN DOORBELL
.Sh SYNOPSIS
.Nm evncr
.Sh DESCRIPTION
EVANNOUNCER, otherwise known as
.Nm ,
is a program which reads the standard input clearly and enjoyably.
.Pp
.Nm
adheres to the UNIX philosophy and is created by VARIK VALEFOR, with
EVAN DOORBELL supplying the voice of
.Nm .
.Pp
.Nm
can read passwords, UIDs, or essentially any form of text data which is
of a reasonable length.
.Pp
.Sh EXAMPLE
Running
.Pp
.Dl echo asshat | evncr
.Pp
should make the terminal speaker output
.Qq LATIN MINUSCULE ALFA, LATIN MINUSCULE SIERRA, LATIN MINUSCULE SIERRA, LATIN MINUSCULE HOTEL, LATIN MINUSCULE ALFA, LATIN MINUSCULE TANGO
in the voice of EVAN DOORBELL.
.Sh FILES
The
.Pa sounds
directory contains the audio files which are played by
.Nm
when
.Nm
reads sequences.  With the exception of pseudocharacters which
are unique to
.Nm ,
e.g.,
.Dq LATIN
and
.Dq MINUSCULE ,
for all printable ASCII characters, the file which contains a character
is named
.Sy [ASCII integer representation of character].wav .
However, majuscule and minuscule characters share source files, with the
phrases
.Dq MAJUSCULE
and
.Dq MINUSCULE
being stored in individual files.  By using this approach, a few bytes
can be saved.  The specifics of the conversion process can be gleaned
from the functions which are called by
.Va parseCharacter .
.Pp
On UNIX-based systems, the
.Pa sounds
folder is located in
.Pa /usr/local/share/evncr/ ,
and on Windows systems, the
.Pa sounds
folder is located in
.Pa [host drive]:\\\\Program Files\\\\EVANNOUNCER\\\\.
.Sh NOTES
As a result of using
.Sy mplayer
as the method of playing sounds, a functional
.Sy mplayer
installation is required.
.Pp
EVAN's voice samples have not yet been provided.
.Pp
Support for different inflections has not yet been implemented.
.Pp
The current cross-platform solution for
.Nm
supports only UNIX-like operating systems.
.Sh HISTORY
The development of the first version of
.Nm
began on 20200709 and lasted two (2) days.  This
.Nm
was abandoned because this
.Nm
was terribly messy and unmaintainable.
.Pp
The previous version of
.Nm
is written in C++ and is a bit unwieldy.  The source code of this
implementation is available at
.Lk http://github.com/varikvalefor/evncr "the EVANNOUNCER GitHub repository" .
.Pp
This version of
.Nm
is written in Haskell and was created because functionality could not be
added to the previous version of
.Nm .
.Sh THANKS
EVAN is a cool dude and supplied the audio files of
.Nm
on a pro-bono basis.
.Pp
This willingness to assist the creation of this silly program is
.Sy very much
appreciated.
.Pp
This section is placed at the end of this document only for the sake of
following tradition.  A big
.Dq THANKS, MAN
would be cool, though.
.Sh CONTACT
Hate mail regarding
.Nm
can be sent to
.An VARIK
.An Dq NOT A COMPUTER PROGRAMMER!
.An VALEFOR
.Aq varikvalefor@aol.com ,
.Em not
to EVAN DOORBELL, who is responsible only for being the mascot and voice
of EVANNOUNCER.


app/evncr.hs
import Config;
import Data.List;
import Control.Monad;
import System.Process;
import Control.Concurrent;

-- | For all 'Character' @k@, @k@ is a representation of an ASCII
-- character.
data Character = Character {
  -- | @prefix1 k@ is the first prefix of @k@.
  --
  -- If @k@ represents a number, then @prefix1 k@ equals @129@.
  -- @prefix1 k@ otherwise equals @128@.
  prefix1 :: Int,
  -- | @prefix2 k@ is the second prefix of @k@, e.g., "MAJUSCULE" or
  -- "MINUSCULE".
  --
  -- @prefix2 k == 0@ iff @k@ is a symbol.
  --
  -- @prefix2 k == 1@ iff @k@ is a numeral.
  --
  -- @prefix2 k == 2@ only if something has gone horribly wrong.
  --
  -- @prefix2 k == 3@ iff @k@ is a majuscule.
  --
  -- @prefix2 k == 4@ iff @k@ is a minuscule.
  prefix2 :: Int,
  -- | @asciiNum k@ is the "meat and potatoes" of @k@.
  --
  -- If @k@ represents a minuscule character, then @asciiNum k@ equals
  -- the ASCII representation of the character which @k@ represents.
  --
  -- If @k@ represents a majuscule character, then @asciiNum k@ equals
  -- the ASCII representation of the minuscule equivalent of the
  -- character which @k@ represents.
  --
  -- If @k@ represents a bracket of some sort, then @asciiNum k@ equals
  -- the ASCII representation of the left bit of @k@'s bracket pair.
  --
  -- If @k@ represents some other symbol, then @asciiNum k@ just equals
  -- the ASCII representation of the character which @k@ represents.
  asciiNum :: Int
} deriving (Eq, Read, Show);

-- | @parseCharacter k@ is a 'Character' which represents @k@.
parseCharacter :: Char -> Character;
parseCharacter k = Character {
  prefix1 = prefix1',
  prefix2 = prefix2',
  asciiNum = asciiNum'
  }
  where
  asciiNum' :: Int
  asciiNum'
    | k' `elem` [97..122] = k' - 32
    | k' `elem` [62,93,125] = k' - 2
    | k' == 41 = 40
    | otherwise = k'
  --
  prefix1' :: Int
  prefix1'
    | k' `elem` [49..58] = 129
    | otherwise = 128
  --
  prefix2' :: Int
  prefix2'
    | k' < 48 = 0
    | k' < 58 = 1
    | k' < 65 = 0
    | k' < 91 = 3
    | k' < 97 = 0
    | k' < 123 = 4
    | otherwise = 0
  --
  k' :: Int
  k' = fromEnum k;

-- | @defaultCharacter@ is the 'Character' value which @'readSingle'@
-- interprets as a "yo, pause for a bit" signal.
defaultCharacter :: Character;
defaultCharacter = Character {prefix1 = 0, prefix2 = 0, asciiNum = 0};

-- | @readSequence k@ reads all characters of @k@.
readSequence :: [Character] -> IO ();
readSequence = mapM_ readSingle . intersperse defaultCharacter;

-- | @readSingle n@ reads @n@.  If @n == 'defaultCharacter'@, then an
-- intercharacter pause is "output".
readSingle :: Character -> IO ();
readSingle k
  | k == defaultCharacter = threadDelay delay_interChar
  | otherwise =
    playFile (prefix1 k) >>
    threadDelay delay_intraChar >>
    playFile (prefix2 k) >>
    threadDelay delay_intraChar >>
    playFile (asciiNum k);

-- | @playFile n@ plays the file whose filename is @show n ++ ".wav"@,
-- where the root directory is @'soundDir'@.
playFile :: Int -> IO ();
playFile n = void $ readProcess "mplayer" [path] []
  where
  path :: FilePath
  path = soundDir ++ show n ++ ".wav";

main :: IO ();
main = getLine >>= readSequence . map parseCharacter;
